### PR TITLE
add proxyconfig control of service-cluster naming scheme

### PR DESCRIPTION
**Pull request by douglas-reid**
Adds control plane implementation of API surface added in https://github.com/istio/api/pull/2174.

This will allow Istio users (mesh admins) to configure how service-cluster names are generated for proxies. This primarily impacts trace span service naming. The PR (with the API PR) is meant to address a regression in control that appeared in 1.12.

Related issue: istio/istio#36162
Prior discussion: istio/istio#12644

- [ X ] Networking
- [ X ] Policies and Telemetry

### CODE REVIEW

File: pkg/bootstrap/config.go [hunk 1]
```diff
@@ -267,20 +268,52 @@ func getLocalityOptions(l *core.Locality) []option.Instance {
...
+		if workloadName == "" {
+			workloadName = "istio-proxy"
+		}
+
+		switch name.TracingServiceName {
+		case meshAPI.ProxyConfig_APP_LABEL_AND_NAMESPACE:
+			return serviceClusterOrDefault("istio-proxy", metadata)
+		case meshAPI.ProxyConfig_CANONICAL_NAME_ONLY:
+			cs, _ := inject.ExtractCanonicalServiceLabels(metadata.Labels, metadata.WorkloadName)
+			return serviceClusterOrDefault(cs, metadata)
```
[the remark this answers is not in the document]
Comment (douglas-reid) [reply pkg/bootstrap/config.go:287]: What are the implications of an empty service cluster name? Is it possible to have an empty `metadata.WorkloadName`? I was worried here about workloads that, for whatever reason, return an empty `cs`.

File: pkg/bootstrap/config.go [hunk 2]
```diff
@@ -593,7 +626,7 @@ func ConvertXDSNodeToNode(node *core.Node) *model.Node {
 	if metadata.ProxyConfig == nil {
 		metadata.ProxyConfig = &model.NodeMetaProxyConfig{}
 	}
-	metadata.ProxyConfig.ServiceCluster = node.Cluster
+	metadata.ProxyConfig.ClusterName = &meshAPI.ProxyConfig_ServiceCluster{ServiceCluster: node.Cluster}
```
[the remark this answers is not in the document]
Comment (douglas-reid) [reply pkg/bootstrap/config.go:629]: this happens when the envoy reports to `istiod`, right?  at which point, the bootstrap setting of service cluster name should have already happened, meaning that `node.Cluster` has the desired string value? My thinking here was that it is no longer important to maintain the distinction, but maybe that is misguided?  I'm not sure how to preserve the format decision through the various transformations if this is not valid.

[the remark this answers is not in the document]
Comment (douglas-reid) [reply pkg/bootstrap/config.go:629]: oh... should this be only in the missing proxyconfig case?

**Comment by douglas-reid** [comment]
@howardjohn ptal. i'm holding off on the istio.io/api dep update, but can do that manually as needed. i've verified that this works when trying to set `CANONICAL_NAME_ONLY`:

image

image

**Comment by douglas-reid** [comment]
@howardjohn ptal. this should now be passing all tests.

File: pkg/bootstrap/config.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (howardjohn) [reply pkg/bootstrap/config.go:286]: Ack, i think this makes sense. There does still seem to be an extremely contrived case where we have `ProxyConfig_CANONICAL_NAME_ONLY`, `service.istio.io/canonical-name=istio-proxy`, and no workloadname or app label set. You would end up with `istio-proxy.namespace` I think...

**Review by howardjohn — APPROVED** [review]

**Review by therealmitchconnors — APPROVED** [review]
approve changes to ux wg files.

File: pilot/cmd/pilot-agent/config/config.go [hunk 3]
```diff
@@ -64,8 +64,10 @@ func ConstructProxyConfig(meshConfigFile, serviceCluster, proxyConfigEnv string,
 		// proxy config.
 		proxyConfig.Concurrency = &types.Int32Value{Value: int32(concurrency)}
 	}
-	if proxyConfig.ServiceCluster == "" {
-		proxyConfig.ServiceCluster = serviceCluster
+	if x, ok := proxyConfig.GetClusterName().(*meshconfig.ProxyConfig_ServiceCluster); ok {
+		if x.ServiceCluster == "" {
+			proxyConfig.ClusterName = &meshconfig.ProxyConfig_ServiceCluster{ServiceCluster: serviceCluster}
```
Comment (hzxuzhonghu) [inline pilot/cmd/pilot-agent/config/config.go:69]:
```suggestion
			x.ServiceCluster = serviceCluster
```

**Comment by zirain** [comment]
fixed by https://github.com/istio/istio/pull/37105

**Pull request merged**